source/Cosmos.HAL2/BlockDevice/Ports/Satapi.cs

```diff
@@ -0,0 +1,170 @@
+
+            FISRegisterH2D xCMDFIS = new FISRegisterH2D(xCMDTable.CFIS)
+            {
+                FISType = (byte)FISType.FIS_Type_RegisterH2D,
+                IsCommand = 1,
+                Command = (byte)ATACommands.Packet,
+                Device = 1 << 4
+            };
+
+            byte[] xATAPICMD = new byte[12];
+            xATAPICMD[0] = (byte)aCommand;
+            xATAPICMD[2] = (byte)((aStart >> 0x18) & 0xFF);
+            xATAPICMD[3] = (byte)((aStart >> 0x10) & 0xFF);
+            xATAPICMD[4] = (byte)((aStart >> 0x08) & 0xFF);
+            xATAPICMD[5] = (byte)((aStart >> 0x00) & 0xFF);
+            xATAPICMD[9] = (byte)(aCount);
+            for (uint i = 0; i < xATAPICMD.Length; i++)
+            new Core.MemoryBlock(xCMDTable.ACMD, 12).Bytes[i] = xATAPICMD[i];
+            
+            int xSpin = 0;
+            do xSpin++; while ((mPortReg.TFD & 0x88) != 0 && xSpin < 1000000);
+
+            if (xSpin == 1000000)
+            {
+                Console.ForegroundColor = ConsoleColor.Red;
+                Console.Write("\n[Error]: ");
+                Console.Write("Port timed out!");
+                Console.ResetColor();
+                return;
+            };
+
+            mPortReg.CI = 1U;
+
+            while(true)
+            {
+                if((mPortReg.CI & (1 << xSlot)) == 0)
+                {
+                    break;
+                }
+                if ((mPortReg.IS & (1 << 30)) != 0)
+                {
+                    Console.ForegroundColor = ConsoleColor.DarkRed;
+                    Console.Write("\n[Fatal]: ");
+                    Console.Write("Fatal error occurred while sending command!");
+                    Console.ResetColor();
+                    return;
+                }
+            }
+
+            if ((mPortReg.IS & (1 << 30)) != 0)
+            {
+                Console.ForegroundColor = ConsoleColor.DarkRed;
+                Console.Write("\n[Fatal]: ");
+                Console.Write("Fatal error occurred while sending command!");
+                Console.ResetColor();
+                return;
+            }
+
+            Console.ForegroundColor = ConsoleColor.Green;
+            Console.Write("\n[Success]: ");
+            Console.Write("Command has been sent successfully!");
+            Console.ResetColor();
+
+            return;
+        }
+
+        public int FindCMDSlot(PortRegisters aPort)
+        {
+            // If not set in SACT and CI, the slot is free
+            var xSlots = (aPort.SACT | aPort.CI);
+
+            for (int i = 1; i < 32; i++)
+            {
+                if ((xSlots & 1) == 0)
+                    return i;
+                xSlots >>= 1;
+            }
+            Console.ForegroundColor = ConsoleColor.Red;
+            Console.Write("\n[Error]: ");
+            Console.Write("Cannot find a free command slot!");
+            Console.ResetColor();
+            return -1;
+        }
```

Comment:
What should happen if CMDSlot is not found? Is this impossible situation again? So...

Reply:
@fanol, no that's not, it happens when all slots are busy, maybe a command is hung inside?

Reply:
OK, I had the doubt it was this. I've not read the SATA specs...